### PR TITLE
[devel] clang-format operations

### PR DESCRIPTION
**Pull request by aMahanna**
### Scope & Purpose

**Objective**
* Introduce a consistent clang-format standard for ArangoDB via Githooks and Actions.

**Implementation**
1.[ Establish a baseline](https://github.com/arangodb/arangodb/pull/15384) by formatting the current codebase via a derivation of the Google style 
(created by @gnusi, see https://github.com/arangodb/arangodb/pull/15131)
2. Introduce a pre-commit githook to automatically format added/modified files locally.
3. Introduce a Github Action workflow to alert on "leftover" clang-format violations

**Baseline example**
* `arangod/**/*` : 78bb583
* `arangosh/**/*`: f4c3b3d
* `lib/**/*`: 8410968
* `tests/**/*`: f0f6424

### CODE REVIEW

**Comment by aMahanna** [comment]
Will establish a clang-format baseline of the codebase upon receiving target directory & version confirmation.

File: scripts/clang-format/clang-format.sh
Comment (jsteemann) [inline scripts/clang-format/clang-format.sh:4]: The assumption of having a directory named `arangodb` feels strange and doesn't work for me right now (and probably won't work for a lot of others).

File: scripts/clang-format/entrypoint.sh
Comment (jsteemann) [inline scripts/clang-format/entrypoint.sh:4]: IMO we don't need to care about `.cu` files here. A common file extension for C++ files is also `.cc`, which is not in the list (and also not used by us). We may or may not want to add this to be "future-proof" :sweat_smile:

File: scripts/clang-format/entrypoint.sh
Comment (jsteemann) [inline scripts/clang-format/entrypoint.sh:2]: The assumption of having to have an `arangodb` directory does not feel right.

File: scripts/clang-format/entrypoint.sh
Comment (jsteemann) [inline scripts/clang-format/entrypoint.sh:6]: If this script is supposed to be run locally by developers, we should also add `enterprise` here, if it exists. So we should test for the existence of the `enterprise` subdirectory, and then apply formatting there too.

File: scripts/clang-format/clang-format.sh
Comment (jsteemann) [reply scripts/clang-format/clang-format.sh:4]: We may get away with something like `pushd ... popd` here. These seem to be supported by most of the relevant shells today.

File: scripts/clang-format/clang-format.sh
Comment (jsteemann) [inline scripts/clang-format/clang-format.sh:2]: I think instead of `${PWD}/scripts/clang-format` we should use `"${PWD}/scripts/clang-format"` here (plus of all following occurrences of variables). Otherwise it may not work with directory names that contain spaces.

**Comment by aMahanna** [comment]
note: PR is on hold until further notice (awaiting publish of the `arangodb/clang-format` image based on https://github.com/aMahanna/arangodb-clang-format/tree/master)

File: .github/workflows/clang-format.yml
Comment (aMahanna) [inline .github/workflows/clang-format.yml:17]: This will be changed to `arangodb/clang-format-action@v1.0.0` once we publish https://github.com/arangodb/clang-format-action

Comment (aMahanna) [reply .github/workflows/clang-format.yml]: Note: CI is failing (https://github.com/arangodb/arangodb/runs/4576063938?check_suite_focus=true) because the ArangoDB Clang-Format Action is yet to be published (see https://github.com/arangodb/clang-format-action)

**Comment by aMahanna** [comment]
Note: https://github.com/arangodb/arangodb/pull/15384 and https://github.com/arangodb/enterprise/pull/831 must be merged before Squashing & Merging this PR

File: scripts/clang-format.sh
Comment (neunhoef) [inline scripts/clang-format.sh:21]: On Linux systems, this will change ownership of all files touched to `root.root`. This would be a nuisance. However, it is easy to provide a second script which runs `clang-format` locally without docker, solving the problem.

**Review by neunhoef — COMMENTED** [review]
I like the fact that the `clang-format.sh` script only looks at changes in comparison to the HEAD. But the file ownership is a (small) problem on Linux.

File: scripts/clang-format.sh
Comment (mpoeter) [reply scripts/clang-format.sh:21]: Running the container with `--user $(id -u):$(id -g)` should avoid this ownership problem.

Comment (aMahanna) [reply scripts/clang-format.sh:21]: I've created a separate [clang-format-linux.sh](https://github.com/arangodb/arangodb/pull/15083/files#diff-fe2859c9145b46a037443925b6569a707229f36e1353d4e227eb06513efa50c6) file to include the changes we discussed.

This script actually works fine on Mac as well (tested it locally). If it happens to work on Windows too, then there will be no need to have 2 different shell scripts

Comment (neunhoef) [reply scripts/clang-format.sh:21]: Yes, I would think so. However, changes are probably not great, but I am no Windows expert.

**Comment by aMahanna** [comment]
OS-dependent hooks & scripts (MacOS, Linux, Windows) have been introduced.

Linux & MacOS implementations are currently identical, but may be subject to change in the future (hence why I still decided to split them up).

**Review by joerg84 — APPROVED** [review]

**Pull request merged**
